olp-cpp-sdk-dataservice-read/src/CatalogClientImpl.cpp

```diff
@@ -124,6 +125,39 @@ CancellableFuture<CatalogVersionResponse> CatalogClientImpl::GetLatestVersion(
   return client::CancellableFuture<CatalogVersionResponse>(
       std::move(cancel_token), std::move(promise));
 }
+
+client::CancellationToken CatalogClientImpl::ListVersions(
+    VersionsRequest request, VersionsResponseCallback callback) {
+  auto schedule_get_versions_list = [&](VersionsRequest request,
+                                        VersionsResponseCallback callback) {
+    auto catalog = catalog_;
+    auto settings = settings_;
+    auto pending_requests = pending_requests_;
+
+    auto versions_list_task =
+        [=](client::CancellationContext context) -> VersionsResponse {
+      return repository::CatalogRepository::GetVersionsList(
+          std::move(catalog), context, std::move(request), std::move(settings));
+    };
+
+    return AddTask(settings.task_scheduler, pending_requests,
+                   std::move(versions_list_task), std::move(callback));
+  };
+
+  return ScheduleFetch(std::move(schedule_get_versions_list),
```

Comment:
Is it even possible to have CacheWithUpdate? As soon as you have the same parameters you always get the same values. I don't think that CacheWithUpdate makes any sense for versioned catalogs.